sparse_strips/vello_cpu/src/dispatch/single_threaded.rs

```diff
@@ -61,13 +111,217 @@ impl SingleThreadedDispatcher {
+                        );
+
+                        fine.pack(region);
+                    });
+
+                    // Apply the filter effect to the completed layer.
+                    fine.filter_layer(&mut pixmap, filter, layer_manager);
+
+                    // Store the filtered result for use by dependent layers.
+                    layer_manager.register_layer(*layer_id, *wtile_bbox, pixmap);
+                }
+                RenderNodeKind::RootLayer {
+                    layer_id,
+                    wtile_bbox: _,
+                } => {
+                    // Final composition directly to output buffer.
+                    let mut regions = Regions::new(width, height, buffer);
+                    regions.update_regions(|region| {
+                        // Use the background color from the wide tile.
+                        let bg = self.wide.get(region.x, region.y).bg;
+                        self.process_layer_tile(
+                            &mut fine,
+                            region.x,
+                            region.y,
+                            *layer_id,
+                            bg,
+                            layer_manager,
+                            encoded_paints,
+                        );
+
+                        assert_eq!(
+                            fine.blend_buf.len(),
+                            1,
+                            "blend buffer should contain exactly one layer after tile processing"
+                        );
+
+                        fine.pack(region);
+                    });
+                }
+            }
+        }
+    }
+
+    /// Processes all rendering commands for a single layer within a specific tile.
+    ///
+    /// This handles the complex logic of composing filtered layers by:
+    /// 1. Running normal rendering commands in sequence.
+    /// 2. When encountering a filtered layer reference, compositing its pre-rendered
+    ///    content from the layer manager.
+    /// 3. Skipping the filtered layer's internal commands (already rendered separately).
+    ///
+    /// # Arguments
+    /// * `fine` - The fine rasterizer instance.
+    /// * `x`, `y` - Wide tile coordinates.
+    /// * `layer_id` - The layer being processed.
+    /// * `clear_color` - Initial color for the tile.
+    /// * `layer_manager` - Storage for filtered layer buffers.
+    /// * `encoded_paints` - Paint definitions for the scene.
+    fn process_layer_tile<S: Simd, F: FineKernel<S>>(
+        &self,
+        fine: &mut Fine<S, F>,
+        x: u16,
+        y: u16,
+        layer_id: u32,
+        clear_color: PremulColor,
+        layer_manager: &mut LayerManager,
+        encoded_paints: &[EncodedPaint],
+    ) {
+        let wtile = self.wide.get(x, y);
+        fine.set_coords(x, y);
+        fine.clear(clear_color);
+
+        // Process all commands in this layer's render range.
+        if let Some(ranges) = wtile.layer_cmd_ranges.get(&layer_id) {
```

Comment:
Can this ever be `None`? Won't it be a logical error if we ever don't have `layer_id`'s commands? It's better to panic on invariants than to fail without knowing about it

Reply:
That’s actually a really good point! Initially, I allocated the full-size pixmap, so some regions didn’t have commands for certain layer_ids, that’s why I needed this check before. But now I see the invariant, and it definitely makes sense to panic here if it’s violated.

Fixed.